### PR TITLE
Optimize the generator to generate a code, matching stricter analysis options

### LINKED ISSUE #43: Please fix: Missing types, consts, final,
Thee are multiple inconsistencies with the pedantic analysis options
The most critical one is implicit-dynamic: false, in the load method, because I have to disable it in order to be able to build with the generated translations.
Here key and value don't have types, they should be at least dynamic, but as the translations should be key/value pairs of strings for a valid json I think you should put types here and also avoid "as" syntax. 
```
_translations = jsonMap
        .map<String, String>((key, value) => MapEntry(key, value as String));
```

Next TranslationKeys
```
class TranslationKeys {
  static final String pleaseAuthenticate = 'please_authenticate';

  static final String viewReadme = 'view_readme';
}
```
The rule that I have is underlining those as 
Prefer const over final for declarations.dart(prefer_const_declarations)

Next Usage of var
```
var jsonStr =
        await rootBundle.loadString('i18n/${locale.languageCode}.json');
```
the rule I have here that is broken is 
Prefer final for variable declarations if they are not reassigned.dart(prefer_final_locals)

Next ```Map jsonMap = json.decode(jsonStr);```
Prefer final for variable declarations if they are not reassigned.dart(prefer_final_locals)

Next ```  String get viewReadme => this.resolve(TranslationKeys.viewReadme);```
Don't access members with `this` unless avoiding shadowing.dart(unnecessary_this)

Next ``` const ['bg', 'en'].contains(locale.languageCode);``` and ```var translation = Translation();```

Specify type annotations.dart(always_specify_types)

I'm sure if this is intended, f the generator actually can make those fixed at all. but generally those are good suggestions to fix, even if you are not using pedantic or w/e analysis options.

### PR DESCRIPTION
**Pull request by talamaska**
fixes https://github.com/TeamWanari/rosetta/issues/43

fix: implicit-dynamic warnings in the generated code
fix: use final instead of vars
add: more typings where finals replace vars
fix: avoid using this where the resolve method is called
change: use static const instead of static final props for the Keys Enum class
add: typed MapEntry

### CODE REVIEW

**Comment by talamaska** [comment]
Any comments?

File: rosetta_generator/lib/src/tree/implementation/visitor.dart [hunk 1]
```diff
@@ -138,9 +138,11 @@ class TranslationVisitor extends Visitor<TranslationProduct, TranslationNode> {
       MethodBuilder methodBuilder, Interceptor interceptor) {
     methodBuilder
       ..type = MethodType.getter
-      ..body = (node.isInHelper ? refThis : refInnerHelper)
-          .property(strResolveMethodName)
+      ..body = refResolve
           .call([keysClassRef.property(node.translation.keyVariable)]).code;
```
Comment (AdamLonger) [inline rosetta_generator/lib/src/tree/implementation/visitor.dart]: This part fails in the generated subclasses:
image

Here we must call the Helper's resolve(...) method, like so:
image

The "this" reference is only redundant in the "Future<void> load(Locale locale)" method, in the Helper.
This part should be updated for this to work:

...
 ..body = (node.isInHelper ? refThis : refInnerHelper)
...

The refThis should be replaced, or the ternary operator reworked .
Also, thank you for the PR! ^^

File: rosetta_generator/lib/src/tree/implementation/visitor.dart [hunk 2]
```diff
@@ -138,9 +138,11 @@ class TranslationVisitor extends Visitor<TranslationProduct, TranslationNode> {
...
+    // ..body = (node.isInHelper ? refThis : refInnerHelper)
+    //     .property(strResolveMethodName + 'test')
+    //     .call([keysClassRef.property(node.translation.keyVariable)]).code;
```
Comment (AdamLonger) [inline rosetta_generator/lib/src/tree/implementation/visitor.dart]: Please remove the unused code

File: rosetta_generator/lib/src/keys.dart [hunk 3]
```diff
@@ -13,7 +13,8 @@ Class generateKeysClass(String keysClassName, List<Translation> translations) {
                   ..name = translation.keyVariable
                   ..type = stringType
                   ..static = true
-                  ..modifier = FieldModifier.final$
+                  // ..modifier = FieldModifier.final$
```
Comment (AdamLonger) [inline rosetta_generator/lib/src/keys.dart]: Please remove the unused code

File: rosetta_generator/lib/src/helper.dart [hunk 4]
```diff
@@ -93,16 +93,19 @@ Method generateLoader(Stone stone, Map<String, Spec> interceptorMap) {
...
             .awaited
-            .assignVar(strLoadJsonStr)
+            .assignFinal(strLoadJsonStr, stringType)
+            // .assignVar(strLoadJsonStr)
             .statement,
         decodeJson
             .call([refJsonStr])
-            .assignVar(strLoadJsonMap, mapOf(stringType, dynamicType))
+            .assignFinal(strLoadJsonMap, mapOf(stringType, dynamicType))
+            // .assignVar(strLoadJsonMap, mapOf(stringType, dynamicType))
```
Comment (AdamLonger) [inline rosetta_generator/lib/src/helper.dart]: Please remove the unused code

File: rosetta_generator/lib/src/helper.dart [hunk 5]
```diff
@@ -93,16 +93,19 @@ Method generateLoader(Stone stone, Map<String, Spec> interceptorMap) {
         assetLoader
             .call([literalString(assetLoaderTemplate)])
             .awaited
-            .assignVar(strLoadJsonStr)
+            .assignFinal(strLoadJsonStr, stringType)
+            // .assignVar(strLoadJsonStr)
```
Comment (AdamLonger) [inline rosetta_generator/lib/src/helper.dart]: Please remove the unused code

File: rosetta_generator/lib/src/delegate.dart [hunk 6]
```diff
@@ -70,7 +70,11 @@ Method _load(String className) {
       ..modifier = MethodModifier.async
       ..requiredParameters.add(localeParameter)
       ..body = Block.of([
-        refer(className).newInstance([]).assignVar(fieldName).statement,
+        refer(className)
+            .newInstance([])
+            .assignFinal(fieldName, typeOf(className))
+            .statement,
+        // refer(className).newInstance([]).assignVar(fieldName).statement,
```
Comment (AdamLonger) [inline rosetta_generator/lib/src/delegate.dart]: Please remove the unused code

**Review by AdamLonger — CHANGES_REQUESTED** [review]

File: rosetta_generator/lib/src/tree/implementation/visitor.dart (hunk 1 above)
Comment (talamaska) [reply rosetta_generator/lib/src/tree/implementation/visitor.dart]: I don't understand, where this class _$Good has come from and why the helper is injected there. The helper is an abstract class, not something you would call. How do I replicate this?

Comment (talamaska) [reply rosetta_generator/lib/src/tree/implementation/visitor.dart]: oh I think I've got it.

Comment (AdamLonger) [reply rosetta_generator/lib/src/tree/implementation/visitor.dart]: Put this in your json:
`{
  "good.luck": "Good luck!",
  "good.idea": "Good idea!"
}`

And this for you translation:
```
part 'translation_grouped_dot_separator.g.dart';

@Stone(path: 'i18n', grouping: Grouping.withSeparator())
class TranslationGroupedDotSeparator
    with _$TranslationGroupedDotSeparatorHelper {
  static LocalizationsDelegate<TranslationGroupedDotSeparator> delegate =
      _$TranslationGroupedDotSeparatorDelegate();

  static TranslationGroupedDotSeparator of(BuildContext context) {
    return Localizations.of(context, TranslationGroupedDotSeparator);
  }

  @Intercept.simple()
  String simpleIntercept(String translation) => translation;
}
```

Long version:
If you put the "good.idea" in the en.json, while a grouping withSeparator() is deffined, then you can reference it like this:
Translation.of(context).good.idea;

This is can be useful if you want to group strings, depending on the screen you are using it (or something).

For this to work Rosetta create the required sub-classes, which can be chained together, to make the above code work.

In the Example's case, it creates a "_$Good" class which hase a "idea" getter, which will return the resolved value, and it creates a _$Good"good" property in the Helper to refrerence it. It's kind of difficult do write down, but if you run the flutter_example you can see it in the generated files :D

Comment (talamaska) [reply rosetta_generator/lib/src/tree/implementation/visitor.dart]: fixed, and tested with the underscore as well.

Comment (AdamLonger) [reply rosetta_generator/lib/src/tree/implementation/visitor.dart]: Nice, thank you!

**Review by AdamLonger — APPROVED** [review]

**Comment by talamaska** [comment]
Any idea when this will be published to pub.dev?

**Comment by AdamLonger** [comment]
Tomorrow, before noon. In around 15 hours from now.

**Comment by AdamLonger** [comment]
rosetta_generator ^0.2.1 is out :)

**Pull request merged**